### PR TITLE
Fifth task

### PR DESCRIPTION
**Pull request by synion**
@kovson

### CODE REVIEW

File: app/controllers/admin/base_controller.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@ class Admin::BaseController < ActionController::Base
   before_action :authenticate_user!
   protect_from_forgery with: :exception
   rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+  after_action :verify_authorized
```
Comment (kovson) [inline app/controllers/admin/base_controller.rb:6]: Why after?

File: app/models/interest.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
 class Interest < ApplicationRecord
-  belongs_to :user
+  self.inheritance_column = :_type_disabled
+  enum type: { health: 0, hobby: 1, work: 2 }
+  belongs_to :user, optional: true
```
Comment (kovson) [inline app/models/interest.rb]: It should be optional?

File: app/models/interest.rb [hunk 3]
```diff
@@ -1,3 +1,5 @@
 class Interest < ApplicationRecord
-  belongs_to :user
+  self.inheritance_column = :_type_disabled
```
Comment (kovson) [inline app/models/interest.rb]: Why we need that?

Comment (synion) [reply app/models/interest.rb]: the type is reserved for rails and this was shut down type for rails and i can now use this in Interest model

File: app/controllers/admin/base_controller.rb (hunk 1 above)
Comment (synion) [reply app/controllers/admin/base_controller.rb:6]: this is from Punditi document say if i forget add authorizaton to action this method remind me on that when not add before

File: app/models/interest.rb (hunk 2 above)
Comment (synion) [reply app/models/interest.rb]: In rails 5.0 i can't create new User with assosiation Interested, i read it is some bug, I can 
do it in configuration new_framework_defaults like a 
Rails.application.config.active_record.belongs_to_required_by_default = false
or ad optional: true, or in User model inverse_of: :user 
But i chose optional :true

File: app/models/interest.rb (hunk 3 above)
Comment (kovson) [reply app/models/interest.rb]: 👍

File: app/controllers/admin/base_controller.rb (hunk 1 above)
Comment (kovson) [reply app/controllers/admin/base_controller.rb:6]: 👍

File: app/models/interest.rb (hunk 2 above)
Comment (kovson) [reply app/models/interest.rb]: If interest have to has ```user_id``` so it is not optional. This option will allow creating ```Interest``` without ```user_id```, it is not what we want to achieve.

File: app/models/user.rb [hunk 4]
```diff
@@ -3,4 +3,11 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
   enum gender: { male: 0, female: 1 }
+  accepts_nested_attributes_for :interests, reject_if: :all_blank, allow_destroy: true
+
+  private
+  def self.create_with_password(attr={})
+    default_password = "secret"
+    self.create(attr.merge(password: default_password, password_confirmation: default_password))
```
Comment (kovson) [inline app/models/user.rb]: I think that you are able to create ```User``` without ```password confirmation```.

File: app/models/user.rb [hunk 5]
```diff
@@ -3,4 +3,11 @@ class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
   enum gender: { male: 0, female: 1 }
+  accepts_nested_attributes_for :interests, reject_if: :all_blank, allow_destroy: true
+
+  private
+  def self.create_with_password(attr={})
```
Comment (kovson) [inline app/models/user.rb]: All users are created with password, maybe ```self.create_by_admin```? Do we need default attributes here?

File: app/models/user.rb [hunk 6]
```diff
@@ -3,4 +3,11 @@ class User < ApplicationRecord
...
+    default_password = "secret"
```
Comment (kovson) [inline app/models/user.rb]: You can just use ```"secret``` as a password in the line below. It is not used many times so there is no need to extract it. 
Consider moving this variable to const.

File: app/models/user.rb (hunk 4 above)
Comment (synion) [reply app/models/user.rb]: Yes you have right

File: app/views/admin/users/_interest_fields.html.haml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+.nested-fields
+  .field
+    = f.label :name
+    %br
+    = f.text_field :name
+  .field
+    =f.label :type
+    =f.select :type, Interest.types.keys
+
+  = link_to_remove_association "remove task", f
```
Comment (kovson) [inline app/views/admin/users/_interest_fields.html.haml]: task?

Comment (synion) [reply app/views/admin/users/_interest_fields.html.haml]: no interest right :)

Comment (synion) [reply app/views/admin/users/_interest_fields.html.haml]: Interest :)

File: spec/features/menage_users_by_admin.rb [hunk 8]
```diff
@@ -38,5 +38,43 @@
...
+
+  scenario 'admin can add new user' do
+    visit admin_users_path
+    expect{ page.to_not have_content("new_user@o2.pl")
+            page.to_not have_content('33')
+            page.to_not have_content("female")
+            page.to_not have_content("Programing")
+            page.to_not have_content("work") }
+    create_user
+    expect{ page.to have_content("new_user@o2.pl")
```
Comment (kovson) [inline spec/features/menage_users_by_admin.rb]: There is one downside to use just one expect. If it fails you don't know where exactly the error is. If you split every single one to: 
```
expect(page).to have_content("new_user@o2.pl")
````
you will know exactly which line fails.

Comment (synion) [reply spec/features/menage_users_by_admin.rb]: Ok this right i will change this

**Comment by kovson** [comment]
@synion 
Few small refactor ideas. Really nice job, I love the test coverage.

File: app/models/user.rb (hunk 5 above)
Comment (synion) [reply app/models/user.rb]: No i was check this second time and you have right i don't need default. When i create this it was't work meaby some small bug ?? :)

File: app/models/interest.rb (hunk 2 above)
Comment (synion) [reply app/models/interest.rb]: @kovson 
Ok but when i want to create User with a interest that give me rollback an user can't create without that  logs : https://gist.github.com/synion/6110d17add32608a071a939607392b52

File: app/controllers/admin/users_controller.rb [hunk 9]
```diff
@@ -39,5 +39,9 @@ def update
   def user_params
     params.require(:user).permit(:email, :age, :gender, interests_attributes:[:id, :name, :type, :_destroy])
   end
+
+  def locals(values)
```
Comment (kovson) [inline app/controllers/admin/users_controller.rb]: Is this ever used?

File: app/controllers/admin/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    end
+  end
+
+  private
+  def user_params
+    params.require(:user).permit(:email, :age, :gender, interests_attributes:[:id, :name, :type, :_destroy])
+  end
+
+  def locals(values)
+  render locals: values
```
Comment (kovson) [inline app/controllers/admin/users_controller.rb]: Is this ever used? I think this method can be removed.

File: app/controllers/admin/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+    authorize :user
+    user = User.find(params[:id])
+    if user.update_attributes(user_params)
+      redirect_to admin_path
+    else
+      render 'edit'
+    end
+  end
+
+  private
```
Comment (kovson) [inline app/controllers/admin/users_controller.rb:38]: One line separation between private and next method declaration.

**Comment by kovson** [comment]
@synion 
Just two small questions.

**Comment by synion** [comment]
@kovson 
I change this i should use local method in action

**Comment by kovson** [comment]
I prefer to use, but your idea is also acceptable. Good job, lets move on :)
```
render locals: { key: value }
```

💃

**Pull request merged**